### ISSUE TITLE
New bug on master

### ISSUE THREAD

**Issue by ilanschnell** [conda/conda#2599]
On master `4.0.8.dev447+89afcdd`, I get:
```
$ CIO_TEST=3 conda update pycrypto
...
Traceback (most recent call last):
  File "/Users/ilan/python/bin/conda", line 6, in <module>
    sys.exit(conda.cli.main.main())
  File "/Users/ilan/conda/conda/cli/main.py", line 120, in main
    args_func(args, p)
  File "/Users/ilan/conda/conda/cli/main.py", line 127, in args_func
    args.func(args, p)
  File "/Users/ilan/conda/conda/cli/main_update.py", line 64, in execute
    install.install(args, parser, 'update')
  File "/Users/ilan/conda/conda/cli/install.py", line 232, in install
    plan.add_defaults_to_specs(r, linked, specs, update=isupdate)
  File "/Users/ilan/conda/conda/plan.py", line 344, in add_defaults_to_specs
    names_linked = {r.index[fn]['name']: fn for fn in linked}
  File "/Users/ilan/conda/conda/plan.py", line 344, in <dictcomp>
    names_linked = {r.index[fn]['name']: fn for fn in linked}
KeyError: 'xar.tar.bz2'
```
@mcg1969

**Comment by mcg1969** [comment]
We can certainly modify like 344 to avoid accessing `r.index` for values of `fn` that are not present. But the `linked` input to `add_defaults_to_specs` is supposed to have a list of the packages installed in that environment. Why does it have `xar` in it? (The previous line is adding the `.tar.bz2`)

**Comment by mcg1969** [comment]
I'm reluctant to make any changes here until we can identify the problem... so if you can give me the `--debug` output when you get back to this, and do any other detective work you can think of, I'd appreciate it.

**Comment by ilanschnell** [comment]
```
$ conda --debug update flask
Using Anaconda Cloud api site https://api.anaconda.org
DEBUG:binstar:Found login token: /Users/ilan/Library/Application Support/binstar/https%3A%2F%2Fapi.anaconda.org.token
DEBUG:conda.fetch:channel_urls=OrderedDict([('https://repo.continuum.io/pkgs/free/osx-64/', ('defaults', 1)), ('https://repo.continuum.io/pkgs/free/noarch/', ('defaults', 1)), ('https://repo.continuum.io/pkgs/pro/osx-64/', ('defaults', 1)), ('https://repo.continuum.io/pkgs/pro/noarch/', ('defaults', 1)), ('https://conda.anaconda.org/t/il-72788b62-7902-4336-99ac-594538e75092/ilan/osx-64/', ('ilan', 2)), ('https://conda.anaconda.org/t/il-72788b62-7902-4336-99ac-594538e75092/ilan/noarch/', ('ilan', 2))])
Fetching package metadata ...INFO:stdoutlog:Fetching package metadata ...
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): repo.continuum.io
DEBUG:requests.packages.urllib3.util.retry:Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): repo.continuum.io
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): repo.continuum.io
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): repo.continuum.io
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): conda.anaconda.org
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): conda.anaconda.org
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/pro/noarch/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/pro/noarch/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1035e8410>}
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/free/osx-64/repodata.json.bz2 HTTP/1.1" 304 0
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/free/noarch/repodata.json.bz2 HTTP/1.1" 304 0
DEBUG:requests.packages.urllib3.connectionpool:"GET /pkgs/pro/osx-64/repodata.json.bz2 HTTP/1.1" 304 0
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/free/noarch/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1035dc210>}
DEBUG:requests.packages.urllib3.connectionpool:"GET /t/il-72788b62-7902-4336-99ac-594538e75092/ilan/noarch/repodata.json.bz2 HTTP/1.1" 200 375
.DEBUG:dotupdate:fetching repodata: args ('https://conda.anaconda.org/t/il-72788b62-7902-4336-99ac-594538e75092/ilan/noarch/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1075f7f90>}
DEBUG:requests.packages.urllib3.connectionpool:"GET /t/il-72788b62-7902-4336-99ac-594538e75092/ilan/osx-64/repodata.json.bz2 HTTP/1.1" 200 1097
.DEBUG:dotupdate:fetching repodata: args ('https://conda.anaconda.org/t/il-72788b62-7902-4336-99ac-594538e75092/ilan/osx-64/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1075f7110>}
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/pro/osx-64/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1035dc990>}
.DEBUG:dotupdate:fetching repodata: args ('https://repo.continuum.io/pkgs/free/osx-64/',) kwargs {'use_cache': False, 'session': <conda.connection.CondaSession object at 0x1035d7dd0>}

INFO:stdoutlog:

DEBUG:conda.plan:H0 specs=['flask']
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/Users/ilan/python/bin/conda", line 6, in <module>
    sys.exit(conda.cli.main())
  File "/Users/ilan/conda/conda/cli/main.py", line 120, in main
    args_func(args, p)
  File "/Users/ilan/conda/conda/cli/main.py", line 127, in args_func
    args.func(args, p)
  File "/Users/ilan/conda/conda/cli/main_update.py", line 64, in execute
    install.install(args, parser, 'update')
  File "/Users/ilan/conda/conda/cli/install.py", line 232, in install
    plan.add_defaults_to_specs(r, linked, specs, update=isupdate)
  File "/Users/ilan/conda/conda/plan.py", line 344, in add_defaults_to_specs
    names_linked = {r.index[fn]['name']: fn for fn in linked}
  File "/Users/ilan/conda/conda/plan.py", line 344, in <dictcomp>
    names_linked = {r.index[fn]['name']: fn for fn in linked}
KeyError: 'xar.tar.bz2'
```

**Comment by mcg1969** [comment]
Argh, there's nothing there that is helping me here. One thing that may help is if you can perhaps send me a tarball of the `conda-meta` directory for that environment. I'm still scratching my head over how `xar` could appear in the list of installed packages.

**Comment by mcg1969** [comment]
Interesting. So you have a file called `xar.json` in your `conda-meta` directory. It apparently represents `xarray-0.7.2-py27_0.tar.bz2`.

**Comment by mcg1969** [comment]
I don't know how `xar.json` got added to your `conda-meta` directory; looking at your history, it *might* have been caused by bugs in 4.1 that we've since quashed. But I would say it is reasonable to fix this one by ignoring any files that are not in the index. I'll create a fix for that.

**Comment by ilanschnell** [comment]
That sounds like a good fix.  Thanks.

**Comment by ilanschnell** [comment]
I can confirm that is if fixed on master (`4.0.8-520-g874e8a7`).  Thanks
